### ISSUE TITLE
New CS proposal: Kubernetes Security Cheatsheet

### ISSUE THREAD

**Issue by ragashreeshekar** [OWASP/CheatSheetSeries#465]
## What is the proposed Cheat Sheet about?
Proposed cheat sheet is about covering the various aspects of security Kubernetes clusters, the available features by Kubernetes itself and some of the common misconfigurations and security best practices to avoid them.

## What security issues are commonly encountered related to this area?
* Use of insecure images
* Using cluster with default configurations which are not secure
* Not restricting/securing access to key cluster components such as ETCD etc
  Reference - https://www.informationsecuritybuzz.com/articles/what-are-the-top-5-kubernetes-security-challenges-and-risks/

## What is the objective of the Cheat Sheet?
Spread the knowledge about good practices when working with Kubernetes.

* Provide information about most common mis-configurations and how to avoid them.
* Inform how good configuration can significantly increase overall application security.

## What other resources exist in this area?
Resources from Kubernetes itself

* https://kubernetes.io/docs/concepts/security/
* https://kubernetes.io/docs/tasks/administer-cluster/securing-a-cluster/

Resources from Google and CNCF

* https://cloud.google.com/kubernetes-engine/docs/how-to/hardening-your-cluster
* https://www.cncf.io/blog/2019/01/14/9-kubernetes-security-best-practices-everyone-must-follow/

External resources

* https://www.stackrox.com/post/2020/05/kubernetes-security-101/

**Comment by mackowski** [comment]
@ragashreeshekar thank you for this proposal. In general, I like the idea but how it will be different than already existing resources? Can you create a short table of contents so we can review what you want to cover there.

**Comment by jmbmxer** [comment]
Hey @ragashreeshekar 👋  I'd be happy to help out with this. I've written and taught quite a bit about the subject and help maintain securekubernetes.com

**Comment by ragashreeshekar** [comment]
Hey team!
Thank you @mackowski. 
Thanks @jmbmxer, that would be very helpful.
I will brainstorm and come up with the table of contents and update here.

**Comment by mackowski** [comment]
Awesome :) I would like to have CS about k8s but it has to bring value and do not duplicate other work. 
E.g. very short CS with points like which k8s distro and/or installers choose and simple rules like enable XYZ + why + simple commends how to check if is enabled + how to enable + link to documentation.
There are services in the cloud for k8s you can also check it.
My idea is that this CS can be an umbrella for other work with simple steps and links to documentation and other resources. But if you come up with sth else I am happy to review and publish

**Comment by c0d3xpl0it** [comment]
Hey Guys, I would like to help on in this CS of K8s but I was just wondering how different would it be from existing project [OWASP Kubernetes Security Testing Guide](https://owasp.org/www-project-kubernetes-security-testing-guide/) ?

**Comment by mackowski** [comment]
In general Testing Guide is more focused on testing/attacking while CheatSheets are meant to be used by defenders, Testing Guide should also go deep in some topics while CheatSheets is aimed to be short reference. 
But as I commented at the beginning please @ragashreeshekar will create ToC and based on that we will decide if this will bring value because we will not be duplicating other work.

**Comment by ragashreeshekar** [comment]
Please find below for the TOC for the K8s security cheat sheet I'm planning to cover. 
The topics are spread across multiple resources and here is a consolidated list. 
As @mackowski rightly suggested, we can have **simple rules like enable XYZ + why + simple commends how to check if is enabled + how to enable + link to documentation.** wherever applicable (I am working on it and will update soon). 
Please let me know your thoughts. Thanks in advance.

---------------------------------------------------------------
# Kubernetes Security Cheatsheet

## Kubernetes Version
Upgrade to the Latest Version
Implement Continuous Security Vulnerability Scanning
Regularly Apply Security Updates to Your Environment

## Limit Direct Access to Kubernetes Nodes
Disable public access
Implement role-based access control
Restrict access to etcd
Controlling access to the Kubelet

## Controlling access to the Kubernetes API 
Use Transport Layer Security (TLS) for all API traffic
API Authentication
API Authorization

## Container images
Ensure That Only Authorized Images are Used in Your Environment
Use minimal base images and avoid adding unnecessary components
Use the latest images/ensure images are up to date. 
Use an image scanner to identify known vulnerabilities

## Pods and Containers
Limit the use of privileged containers
Assess the privileges used by containers
Use labels and annotations appropriately
Monitor network traffic to limit unnecessary or insecure communication
Compare and analyze different runtime activity in pods of the same deployments
Use Kubernetes built-in controls when available to tighten security
Apply Security Context to Your Pods and Containers

## Securing data
Keeps secrets as secrets
Restrict access to secrets 
Encrypt secrets at rest

## Controlling the capabilities of a workload or user at runtime 
Limiting resource usage on a cluster
Controlling what privileges containers run with
Preventing containers from loading unwanted kernel modules
Restricting network access
Controlling which nodes pods may access

## Protecting cluster components from compromise
Rotate infrastructure credentials frequently
Encrypt secrets at rest
Receiving alerts for security updates and reporting vulnerabilities

## Log everything
Enable audit logging
Define Audit Policies

## Final thoughts
Embed security earlier into the container lifecycle
Use Kubernetes-native security controls to reduce operational risk

Leverage the context that Kubernetes provides to prioritize remediation efforts

**Comment by mackowski** [comment]
Looks good for me. You can start working on it.
@jmbmxer @c0d3xpl0it - what is your opinion? Do you want to add anything?

**Comment by Seppl2202** [comment]
Hi all,
I did my bachelor thesis on security in K8s and Docker, maybe I can help out with some of the sections, just tell me :)

**Comment by ragashree-mc** [comment]
Thanks @mackowski. I have started working on it, first draft version would be available probably by end of week. 
Thank you @Seppl2202 @jmbmxer @c0d3xpl0it for offering your time and experience. Request you to review the TOC and provide your review comments based on which we can develop the cheatsheet accordingly.

**Comment by Seppl2202** [comment]
> Thanks @mackowski. I have started working on it, first draft version would be available probably by end of week.
> Thank you @Seppl2202 @jmbmxer @c0d3xpl0it for offering your time and experience. Request you to review the TOC and provide your review comments based on which we can develop the cheatsheet accordingly.

That sounds great!

**Comment by c0d3xpl0it** [comment]
@ragashreeshekar @ragashree-mc the provided TOC is enough to start the cheatsheet preparation. We can start adding gradually to same later.

**Comment by ragashreeshekar** [comment]
I have created the initial draft and wanted to share the same. 
Wanted to understand the next steps, would you want me to create a PR for the same linking to this issue or do I post the content here? (It's quite huge). Thanks in advance. 

Regards,
Ragashree

**Comment by mackowski** [comment]
@Seppl2202 @jmbmxer @c0d3xpl0it @ragashree-mc Please review #470 :-)

**Comment by jmbmxer** [comment]
Just submitted some feedback and additions. Can we get this into a draft format so we can begin iterating? What is the next step?

**Comment by ragashreeshekar** [comment]
Thanks @jmbmxer @Seppl2202  @c0d3xpl0it for your review and suggestions, it's been very helpful. 
I have incorporated the information at hand to the cheatsheet, will continue to research on the remainder of them and update it as and when I have the information.

**Comment by mackowski** [comment]
When you will be happy about draft we can add it to 'cheatsheets_draft' directory and start adding there smaller PR and iterate it and finally move it to regular CSs. After we close this initial PR and draft will be in 'cheatsheets_draft' you will be able to open PRs independently and work in parallel.

**Comment by mackowski** [comment]
@ragashreeshekar @Seppl2202 @jmbmxer @c0d3xpl0it @ragashree-mc - are you still working on PRs to the draft CS? Do you need any help?

**Comment by mackowski** [comment]
Nice! Thank you for the update.

**Comment by ragashreeshekar** [comment]
I have added content (PR 496 - https://github.com/OWASP/CheatSheetSeries/pull/496) for OPA and Service Mesh to the best of my understanding and exposure. 
As far as I understand, we have already covered the "Managed Kubernetes Security Considerations" across the entirety of this CS. I request the review of the same and please let us know if we need to update with any changes. Thanks in advance.

**Comment by ragashreeshekar** [comment]
Greetings! Please let us know the next steps for the K8s security CS PR.

**Comment by mackowski** [comment]
@ragashreeshekar apologies I missed that PR. I done review. I like it very much. Please review my comments in the PR. 
@Seppl2202 @jmbmxer @c0d3xpl0it @ragashree-mc  - can you do review of the PR 496 - #496 before merge?

**Comment by Seppl2202** [comment]
@mackowski Sorry for beeing inactive the last weeks, I was on vacation. I will review the PR

**Comment by ragashreeshekar** [comment]
Thank you Team. Thanks @mackowski 
@mackowski Now that the CS in merged and present in the CS drafts, request your thoughts on how we need to go about so it gets included in the CS folder. Please advise.

**Comment by mackowski** [comment]
@ragashreeshekar apologies for late replay! If you think that this is ready please make a PR that moves this CS from draft to regular folder. We should release it ASAP ;-)

**Comment by ragashreeshekar** [comment]
Thanks @mackowski and the entire community for all your guidance and reviews. 
I have moved it to cheatsheets folder pls find the same in PR #525 https://github.com/OWASP/CheatSheetSeries/pull/525.
Please take it forward.
Regards, Raga

**Comment by mackowski** [comment]
The CS should be live now :) Thank you all!
